src/sagas/walletConnect.js

```diff
@@ -0,0 +1,494 @@
+import {
+  WALLET_CONNECT_PROJECT_ID,
+  WALLET_CONNECT_FEATURE_TOGGLE,
+  NETWORK,
+} from '../constants';
+import {
+  types,
+  setWalletConnect,
+  setWalletConnectModal,
+  setWalletConnectSessions,
+  onExceptionCaptured,
+  setWCConnectionFailed,
+} from '../actions';
+import { checkForFeatureFlag, showPinScreenForResult } from './helpers';
+
+const AVAILABLE_METHODS = {
+  HATHOR_SIGN_MESSAGE: 'hathor_signMessage',
+};
+const AVAILABLE_EVENTS = [];
+
+/**
+ * Those are the only ones we are currently using, extracted from
+ * https://docs.walletconnect.com/2.0/specs/clients/sign/error-codes
+ */
+const ERROR_CODES = {
+  UNAUTHORIZED_METHODS: 3001,
+  USER_DISCONNECTED: 6000,
+  USER_REJECTED: 5000,
+  USER_REJECTED_METHOD: 5002,
+  INVALID_PAYLOAD: 5003,
+};
+
+function* isWalletConnectEnabled() {
+  const walletConnectEnabled = yield call(checkForFeatureFlag, WALLET_CONNECT_FEATURE_TOGGLE);
+
+  return walletConnectEnabled;
+}
+
+function* init() {
+  try {
+    const walletConnectEnabled = yield call(isWalletConnectEnabled);
+
+    if (!walletConnectEnabled) {
+      return;
+    }
+
+    const core = new Core({
+      projectId: WALLET_CONNECT_PROJECT_ID,
+    });
+
+    const metadata = {
+      name: 'Hathor',
+      description: 'Hathor Mobile Wallet',
+      url: 'https://hathor.network/',
+    };
+
+    const web3wallet = yield call(Web3Wallet.init, {
+      core,
+      metadata,
+    });
+
+    yield put(setWalletConnect({
+      web3wallet,
+      core,
+    }));
+
+    yield fork(setupListeners, web3wallet);
+
+    // Refresh redux with the active sessions, loaded from storage
+    yield call(refreshActiveSessions);
+
+    // If the wallet is reset, we should cancel all listeners
+    yield take(types.RESET_WALLET);
+
+    yield cancel();
+  } catch (error) {
+    console.error('Error loading wallet connect', error);
+    yield put(onExceptionCaptured(error));
+  }
+}
+
+export function* refreshActiveSessions() {
+  const { web3wallet } = yield select((state) => state.walletConnect.client);
+
+  const activeSessions = yield call(() => web3wallet.getActiveSessions());
+  yield put(setWalletConnectSessions(activeSessions));
+}
+
+/**
+ * @param {Web3Wallet} web3wallet The WalletConnect web3wallet instance
+ */
+export function* setupListeners(web3wallet) {
+  const channel = eventChannel((emitter) => {
+    const listener = (state) => emitter(state);
+
+    web3wallet.on('session_request', (event) => {
+      emitter({
+        type: 'WC_SESSION_REQUEST',
+        data: event,
+      });
+    });
+
+    web3wallet.on('session_proposal', async (proposal) => {
+      emitter({
+        type: 'WC_SESSION_PROPOSAL',
+        data: proposal,
+      });
+    });
+
+    web3wallet.on('session_delete', async (session) => {
+      emitter({
+        type: 'WC_SESSION_DELETE',
+        data: session,
+      });
+    });
+
+    return () => {
+      web3wallet.removeListener('session_request', listener);
```

Comment:
Does it work like this? Because you create the `listener` variable but you don't use it in the `on` events.

Reply:
Fixed and improved in b4cbef7d3ce8f0a60f52f9db3d91f9523e69da1e